### DISCUSSION TITLE
New Architecture Known Issues & Troubleshooting

### DISCUSSION THREAD

**Discussion by cortinico** [reactwg/react-native-new-architecture#237] [Documentation]
**tl;dr: This post collects all the known issues with related resolutions to help users migrate the the New Architecture** 

Hi all,

With React Native 0.76 approaching, we want to collect in a single place all the known issues and troubleshooting we collected from the previous months of experimentation with the New Architecture.

This is a living document and we’ll be updating this document as we get more bug reports. If you encounter any issues migrating to the New Architecture, please open a new issue [using this template](https://github.com/facebook/react-native/issues/new?assignees=&labels=Needs%3A+Triage+%3Amag%3A%2CType%3A+New+Architecture&projects=&template=new_architecture_bug_report.yml).

### Unable to use `Object.keys` on NativeModules

**Answer**: This is by design. We designed the TurboModule system with the goal of lazy loading, so we don’t have to load all the modules in memory when the app is launched. This should let your app start faster. 

One of the trade-offs of this solution is that you cannot call `Object.keys` on a native module right away, because not all the keys might be available immediately. 

Instead, you can read them by leveraging the inheritance. You can call `Object.keys` on the native module prototype and the result should be the same as before. 
```diff
- for prop in Object.keys(nativeModule) {
+ for prop in Object.keys(Object.getPrototypeOf(nativeModule)) {
 // use the prop here
}
```

** [iOS] Cannot access `CallInvoker` from `RCTCxxBridge.callInvoker`

**Answer**: In bridgeless mode, the bridge is not initialized anymore. So, if you were retrieving the `jsCallInvoker` instance from the `[RCTBridge currentBridge]` invocation, that might not work well anymore.

If you need this functionality, the best approach is to implement a pure C++ TurboModule which receives the callInvoker as a parameter when it is initialized. 

This will also help you write more encapsulated code.

## Components Known Issues

** Components with custom Shadow Nodes and measurement are not working with the Interop Layer.

Answer: This is by design because the New Architecture stores the shadow node and layout information in the C++ layer. If your custom component/library is providing a custom shadow node implementation you won’t be able to use it via the Interop Layer (this is a known limitation of the Interop Layer).

Specifically on Android, if your `ViewManager` class is implementing any of those methods:

* `createShadowNodeInstance()` ([reference](https://github.com/facebook/react-native/blob/723a37cb26d125b545d49a29639c27c965a306e2/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManager.java#L135-L142))
* `createShadowNodeInstance(ReactApplication) `([reference](https://github.com/facebook/react-native/blob/723a37cb26d125b545d49a29639c27c965a306e2/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManager.java#L144-L146))
* `createShadowNodeClass()` ([reference](https://github.com/facebook/react-native/blob/723a37cb26d125b545d49a29639c27c965a306e2/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManager.java#L148-L158))
* `measure(Context,ReadableMap,ReadableMap,ReadableMap,float,YogaMeasureMode, float, YogaMeasureMode heightMode, float[]) `([reference](https://github.com/facebook/react-native/blob/723a37cb26d125b545d49a29639c27c965a306e2/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManager.java#L144-L146))

Your component won’t be working with the Interop Layer and will require an implementation with the new Renderer.

**Comment by myselfuser1** [comment]
Can you please check [this](https://github.com/facebook/react-native/issues/47217)

**Comment by sidorchukandrew** [comment]
Does this mean `createShadowNodeInstance`,  `createShadowNodeClass` and `measure` _do_ work with Fabric on Android? I created a new library with builder bob but `measure` and `createShadowNodeInstance` are never called. I'm 99% sure I do _not_ have the interop layer on because my logs say `Bridgeless mode is enabled` and `fabric: true`. The component renders to the screen but is not clickable because it doesn't have a height and width.

**Reply by cortinico** [reply]
> Does this mean createShadowNodeInstance, createShadowNodeClass and measure do work with Fabric on Android?

As mentioned in the post, nope your component won't work. What's happening in your case is that the Android native view is created and added on screen, but those methods are never called because Fabric is expecting layout information coming from the C++ layer, so your view will be unusable

**Reply by sidorchukandrew** [reply]
Thank you for the explanation! Is there documentation on how to make it work through the C++ layer? I'm playing around with the community RN slider package and I also see `Bridgeless mode is enabled` and `fabric: true` in the console logs. When I drop a breakpoint [right here](https://github.com/callstack/react-native-slider/blob/main/package/android/src/newarch/java/com/reactnativecommunity/slider/ReactSliderManager.java#L216), the breakpoint is actually hit. If I understand your most recent comment, this shouldn't be the case because the measurements should be coming from C++. 

It's obvious that I'm misunderstanding something, so any clarifications would be very appreciated!

**Reply by sidorchukandrew** [reply]
Also thank you for such a quick response! That is much appreciated!

> ** Components with custom Shadow Nodes and measurement are not working with the Interop Layer.

Reading this, I thought that meant that if the Interop layer is _on_, then the measurements won't work. But if it's off, measuring would continue to work. I also thought the interop layer only turns on when a component does not support the new architecture. My intention is to create a component that _only_ supports the new architecture, so the interop layer should be _off_.

**Reply by sidorchukandrew** [reply]
I just found [this PR](https://github.com/callstack/react-native-slider/pull/589) that shows how the slider uses C++ to measure!

**Reply by sidorchukandrew** [reply]
Since Fabric is the future, documentation on how to implement these shadow modes and measurements, and how to connect it with the library would be immensely helpful! I'm almost tempted to only use the old architecture for now because it's very difficult to determine what the right steps are. Regardless, this is all very exciting stuff!

**Reply by cortinico** [reply]
Nope we don't have custom C++ documentation at the moment as that's sort of an advanced topic. Documentation is extremely costly to write and can get stale really easy so we focused on the most common scenarios first. As a remediation, you can check how the core components like `Switch` and others integrate as they all go through custom C++ code.

We hope to have capacity to add further New Architecture documentation in the near future

**Comment by HemantRMali** [comment]
Hi @cipolleschi  and @cortinico,


It has been quite a while since we've been trying to figure out the issue with redux-persist specially with persistStore.

Interestingly, when we run the application in debug mode during development, we never encounter the issue. However, after providing the build to QA following the RN migration, we are experiencing crashes on iOS without any logs.

Here is the store file that we are using in our application:


import { configureStore } from '@reduxjs/toolkit';
import { persistReducer, Persistor, persistStore, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import rootReducer from './rootReducer';
import middlewares, { sagaMiddleware } from './middlewares';
import createPersistConfig from './persistConfig';
import reactotron from '@appName/ReactotronConfig';
import rootSaga from './rootSaga';

import { Store } from 'redux';

let store: Store;
let persistor: Persistor;

const enhancers = (getDefaultEnhancers: any) => {
  if (__DEV__) {
    return getDefaultEnhancers().concat(reactotron.createEnhancer());
  }
  return getDefaultEnhancers();
};

export const generateStore = (secretKey: string | null) => {
  // @ts-ignore
  const reducer = persistReducer(createPersistConfig(secretKey), rootReducer);
  store = configureStore({
    reducer,
    enhancers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        },
        immutableCheck: false,
      }).concat(middlewares),
  });

  persistor = persistStore(store); => WE NOTICED A CRASH ON THIS LINE POST MULTIPLE TRIALS AND ERRORS.

  sagaMiddleware.run(rootSaga);

  return store;
};

export { store, persistor };


Below are the package details that we are using related to redux and react native version:

  "dependencies": {
    "@reduxjs/toolkit": "^2.1.0",
    "react": "18.3.1",
    "react-native": "0.77.1",
    "react-redux": "^9.1.0",
    "redux": "4.2.1",
    "redux-axios-middleware": "^4.0.1",
    "redux-persist": "^6.0.0",
    "redux-remember": "^5.2.0",
    "redux-saga": "^1.2.3",
  },

"devDependencies": {
    "@redux-saga/testing-utils": "^1.1.5",
    "reactotron-react-native": "5.1.7",
    "reactotron-redux": "3.1.9",
  },

We appreciate any support from the community to help fix this issue.

**Reply by cortinico** [reply]
> we are experiencing crashes on iOS without any logs.

How are we supposed to help you without reproducers and logs?

**Reply by HemantRMali** [reply]
Hi @cortinico  Thanks for your response...

Here is the stack trace, please check. 


libc++abi: terminating due to uncaught exception of type facebook::jsi::JSError: Exception in HostFunction: -[NSNull stringByTrimmingCharactersInSet:]: unrecognized selector sent to instance 0x20b71b910

Error: Exception in HostFunction: -[NSNull stringByTrimmingCharactersInSet:]: unrecognized selector sent to instance 0x20b71b910
    at requireModule (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:2954:56)
    at getEnforcing (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:2965:31)
    at anonymous (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:44602:68)
    at loadModuleImplementation (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:158:14)
    at guardedLoadModule (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:96:38)
    at metroRequire (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:38:92)
    at anonymous (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:44578:60)
    at loadModuleImplementation (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:158:14)
    at guardedLoadModule (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:96:38)
    at metroRequire (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:38:92)
    at anonymous (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:44496:59)
    at loadModuleImplementation (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:158:14)
    at guardedLoadModule (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:89:47)
    at metroRequire (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:38:92)
    at get AppState (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:1850:25)
    at _default (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:269026:54)
    at _default (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:269091:50)
    at _default (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:263452:81)
    at renderWithHooks (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:5241:24)
    at beginWork (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:6533:36)
    at performUnitOfWork (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:8109:25)
    at workLoopSync (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:8020:55)
    at renderRootSync (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:8004:19)
    at performConcurrentWorkOnRoot (/Users/user_name/Library/Developer/Xcode/DerivedData/app-hccabrmrzspimofqfjaylkbeeczn/Build/Products/Release-iphoneos/main.jsbundle:7811:160)

![Screenshot 2025-03-21 at 3 36 12 PM](https://github.com/user-attachments/assets/23ed395b-1940-4916-902d-d972a005161d)
![Screenshot 2025-03-21 at 3 17 28 PM](https://github.com/user-attachments/assets/1ba1aa77-e38c-4ce2-b35a-840e96f27a2e)




@cipolleschi

**Reply by HemantRMali** [reply]
Here is stack trace from Crashlytics.
@cortinico  @cipolleschi 
[com.mobile_issue_e98c904cb8a5e12619221f1b05a6_crash_session_1990b1240e9e4e748fce64d3efe9_DNE_0_v2_stacktrace.txt](https://github.com/user-attachments/files/19386501/com.mobile_issue_e98c904cb8a5e12619221f1b05a6_crash_session_1990b1240e9e4e748fce64d3efe9_DNE_0_v2_stacktrace.txt)

**Reply by cortinico** [reply]
Can we get a reproducer by any chance?

**Reply by HemantRMali** [reply]
Hi @cortinico 

We just figured out the root cause of the issue and wanted to share it here in case anyone else encounters the same scenario.

In the release debug, we noticed the following crash:

```
libc++abi: terminating due to uncaught exception of type facebook::jsi::JSError: Exception in HostFunction: -[NSNull stringByTrimmingCharactersInSet:]: unrecognized selector sent to instance 0x20b71b910
```

The root cause was this line: 

```javascript
UrbanAirship.setNamedUser(email ? md5(email.toLowerCase()) : null);

Possibly new architecture requires strict type checking..???
```

A `null` value was being passed because the method has the following type options:

```typescript
static setNamedUser(namedUser: string | null | undefined): void;
```

However, when we checked this Native Module on the iOS side, it was expecting an `NSString`, as shown below:

```objc
RCT_EXPORT_METHOD(setNamedUser:(NSString *)namedUser) {
    // rest of the code
    namedUser = [namedUser stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    // rest of the code
}
```

Since it required a string, I changed it on the React Native side to the following, and it started working without any crashes:

```javascript
UrbanAirship.setNamedUser(email ? md5(email.toLowerCase()) : "");
```

I changed it to an empty string, and it worked.

Possibly, the new architecture requires stricter type checking?

cc: @cipolleschi

**Reply by HemantRMali** [reply]
> Hi @cortinico
> 
> We just figured out the root cause of the issue and wanted to share it here in case anyone else encounters the same scenario.
> 
> In the release debug, we noticed the following crash:
> 
> ```
> libc++abi: terminating due to uncaught exception of type facebook::jsi::JSError: Exception in HostFunction: -[NSNull stringByTrimmingCharactersInSet:]: unrecognized selector sent to instance 0x20b71b910
> ```
> 
> The root cause was this line:
> 
> ```js
> UrbanAirship.setNamedUser(email ? md5(email.toLowerCase()) : null);
> 
> Possibly new architecture requires strict type checking..???
> ```
> 
> A `null` value was being passed because the method has the following type options:
> 
> ```ts
> static setNamedUser(namedUser: string | null | undefined): void;
> ```
> 
> However, when we checked this Native Module on the iOS side, it was expecting an `NSString`, as shown below:
> 
> ```objc
> RCT_EXPORT_METHOD(setNamedUser:(NSString *)namedUser) {
>     // rest of the code
>     namedUser = [namedUser stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
>     // rest of the code
> }
> ```
> 
> Since it required a string, I changed it on the React Native side to the following, and it started working without any crashes:
> 
> ```js
> UrbanAirship.setNamedUser(email ? md5(email.toLowerCase()) : "");
> ```
> 
> I changed it to an empty string, and it worked.
> 
> Possibly, the new architecture requires stricter type checking?
> 
> cc: @cipolleschi

I have to consider using https://github.com/urbanairship/react-native-airship instead of older versions of airship.

its     "urbanairship-react-native": "^14.7.0" in my project. (Man shrug)

**Reply by cipolleschi** [reply]
@HemantRMali  Yeah, the New Architecture has built on the assumption that Codegen will ensure typesafety. Of course, for a module going through the interop layer, that can't be ensured. Recently, I made some changes in the interop layers to make sure that the old arch semantic is respected in the New Arch, while using Interop Layers.

On which version were you testing?

**Reply by HemantRMali** [reply]
Hi @cipolleschi , Thanks for your reply.

I am on 0.77.1 version.

**Reply by cipolleschi** [reply]
Can you try bumping to 0.77.2? I think I fixed already those issues.
